### PR TITLE
alloc: fix potential oob in rrealloc, rbrealloc

### PR DESCRIPTION
**Pull request by mmaka1**
Previous memcpy of bytes is unsafe if new object is larger
then old one (which is probably most popular realloc use).

Typically realloc retrieves the old size internally and does
not need this information passed as argument but sof heap
implementation does not store the exact size of the allocated
memory buffer.

Signed-off-by: Marcin Maka <marcin.maka@linux.intel.com>

### CODE REVIEW

**Comment by mmaka1** [comment]
Another fix is to remove this function completely or improve it for constrained environment where temporary double allocation of large objects might simply not work. Note that #2730 removes the only known use of this API.

File: src/include/sof/lib/alloc.h [hunk 1]
```diff
@@ -111,13 +111,14 @@ static inline void *rballoc(uint32_t flags, uint32_t caps, size_t bytes)
  * @param flags Flags, see SOF_MEM_FLAG_...
  * @param caps Capabilities, see SOF_MEM_CAPS_...
  * @param bytes New size in bytes.
+ * @parma old_bytes Old size in bytes.
  * @return Pointer to the resized memory or NULL if failed.
```
Comment (dbaluta) [inline src/include/sof/lib/alloc.h:118]: typo parma

Comment (mmaka1) [reply src/include/sof/lib/alloc.h:118]: Fixed

File: src/lib/alloc.c [hunk 2]
```diff
@@ -913,11 +913,12 @@ void rfree(void *ptr)
 }
 
 void *rrealloc(void *ptr, enum mem_zone zone, uint32_t flags, uint32_t caps,
-	       size_t bytes)
+	       size_t bytes, size_t old_bytes)
 {
 	struct mm *memmap = memmap_get();
 	void *new_ptr = NULL;
 	uint32_t lock_flags;
+	size_t copy_bytes = bytes < old_bytes ? bytes : old_bytes;
```
Comment (jajanusz) [inline src/lib/alloc.c:921]: Would go for just `min()` macro

Comment (mmaka1) [reply src/lib/alloc.c]: Done

**Comment by mmaka1** [comment]
There will be another patch to remove rrealloc from the code once #2730 (and this one to have fixed version archived just in case) are merged.

**Comment by lgirdwood** [comment]
I've tagged for v1.5, so will recheck after the final patch is added.

**Comment by mmaka1** [comment]
> I've tagged for v1.5, so will recheck after the final patch is added.

@lgirdwood I do not plan any other patch in this PR. I'd merge fixed version and then submit another PR to remove rballoc since this action depends on #2730 too.

**Pull request merged**